### DISCUSSION TITLE
Do not require fixing a fixed field of a ref struct

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#9014] [Language Ideas]
From #1502 

A fixed statement should not be required on a fixed field of a ref struct, because it is never moveable.

Note for future LDM consideration: there is a bug tail involved with any `fixed` change. While this is probably not worth it on it's own merits, if we have other `fixed` work that result in refactoring this would be worth it at that point.

**Comment by 4creators** [comment]
Encountered this problem multiple times while working with HW intrinsics.

**Comment by jaredpar** [comment]
This is very related to #1697

**Comment by tannergooding** [comment]
Should this just be, don't requiring pinning when taking the address of a `ref struct` or its fields?

The entire `ref struct` is fixed and should be safe to take the address of (not just `this` and not just `fixed` fields).

**Comment by gafter** [comment]
@tannergooding I think you’re right

**Comment by ltrzesniewski** [comment]
@tannergooding I didn't talk about fields in #1697 but I probably should have. I think it should be safe to take both the address of the `ref struct` and of its unmanaged fields.

**Comment by kkokosa** [comment]
It seems I've already proposed a more general question regarding all fields in #1564

**Comment by tannergooding** [comment]
@gafter, does this need to go to LDM or is it already covered by the languge spec?

The fix, overall, looks rather trivial. From what I was able to determine, this ultimately comes down to what we return from `IsMoveableVariable` (http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Binder_Operators.cs,bfd68ac15f2dece4) and the fix would be to check if `expr.Type.IsRefLikeType` (or a similar check, depending on the expression kind)....

**Comment by tannergooding** [comment]
I guess this does need to go to LDM, since it could be a breaking change...

Normally, the compiler reports `ERR_FixedNotNeeded ` or `ERR_FixedNeeded`. `ref structs` should fall under the former, but currently are treated as the latter.

If this was simply changed to not require `fixed`, than any existing code which is fixing a `ref struct` would now fail to compile as `ERR_FixedNotNeeded` would be reported. So, it would need to be that:
* We take a breaking change and have people fix up their code
* We don't take the change because it is breaking
* We take the change, but allow fixing or not so that existing code continues to work
* Some other alternative not listed

**Comment by tannergooding** [comment]
I've prototyped something that looks to work here: https://github.com/tannergooding/roslyn/commit/f31dbcd5ffa5aaaf91cfcd8b485379dc898168ae

It goes with the first option (which is most likely not what LDM will decide given the compat bar) and I'm sure I missed some edge cases, but I'll add more tests to find out.

**Comment by PathogenDavid** [comment]
> * We take the change, but allow fixing or not so that existing code continues to work

I think this solution along with adding a warning to a warning wave might be the best solution to add this without breaking compatibility. (An error would be ideal, but [LDM decided against error waves](https://github.com/dotnet/csharplang/blob/b6872068faa1b34ae3c9585d182908729b10aae4/meetings/2018/LDM-2018-09-26.md#q--a).)

That and also probably omit pinned flag on the corresponding local when the unnecessary fixed statement is used.

**Comment by jaredpar** [comment]
> We take a breaking change and have people fix up their code

Don't think that is a realistic option here. Code which is `unsafe` is more likely to be legacy than not. There is plenty in Visual Studio for example that hasn't been touched in a decade. Asking developers to change all of that up, particularly because it's an area which can have very subtle bugs, just to upgrade the compiler is not realistic.

> We take the change, but allow fixing or not so that existing code continues to work

Much better 😄

**Comment by tannergooding** [comment]
Just commenting here as I always need to go looking for it again.

The LDM last reviewed this here: https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-10-14.md#do-not-require-fixing-a-fixed-field-of-a-ref-struct and its on the long term backlog

**Comment by ltrzesniewski** [comment]
@jaredpar The issue has been closed without comment, converted into this discussion and #9015 has been opened then immediately locked. Could we get an update on the status of this proposal?
